Sibyll*

afedynitch opened the pull request:
This PR combines multiple items (sorry for that). The main novelty is SIBYLL* as shown at ICRC by F. Riehn:

1. Include SIBYLL* (thx @jncots)
2. When using Ninja for compilation, the FORTRAN files were pre-processed twice, once due to the `-cpp` flag in CMake, and a second time by Ninja itself that calls the preprocessor by default. Fixing this restores the quick build times on machines with many cores.
3. The allowed projectiles for all of the models have been reviews and the lists extended where appropriate
4. Some fixes to UrQMD file list (exclude the includes from f2py)

Opinions desired on:

1. If the SIBYLL* subclasses should be called using the underscore or camel case for consistency?
2. If I should put effort in using only one source file for SIBYLL2.3d and SIBYLL*. In principle, one can restore original SIbyll2.3d behaviour from the sibyll* source file.
 

jncots commented:
I think that:

* The name of classes should be changed standard camel style, like `Sibyll23dStarNoEnh`
* There should be only 1 file to avoid code duplication.

HDembinski commented on src/chromo/models/dpmjetIII.py:
```diff
@@ -213,10 +213,11 @@ def _generate(self):
 
 class DpmjetIII191(DpmjetIIIRun):
     _version = "19.1"
+    _projectiles = standard_projectiles | Nuclei() | {3322, 3312, 3222, 3122, 3112, 311}
```
If it is shared between DpmjetIII191 and  DpmjetIII193, it should probably be in DpmjetIIIRun. This also works, but putting common code there is cleaner.

HDembinski commented on src/chromo/models/dpmjetIII.py:
```diff
@@ -213,10 +213,11 @@ def _generate(self):
     _library_name = "_dpmjetIII191"
 
 
-class DpmjetIII193(DpmjetIIIRun):
+class DpmjetIII193(DpmjetIII191):
```
If you put the extra code into DpmjetIIIRun, you can revert this change.

afedynitch replied:
ok

HDembinski commented on src/chromo/models/sibyll.py:
```diff
@@ -137,34 +141,80 @@ class Sibyll23(SIBYLLRun):
 
 class Sibyll23c(SIBYLLRun):
     _version = "2.3c"
+    _projectiles = standard_projectiles | {
+        3112,
+        3122,
+        3312,
+        3322,
+        3222,
+        411,
+        421,
+        4232,
+        431,
+        4122,
+        4132,
+        4232,
+        431,
+        4332,
+    }
```
Likewise, shouldn't this be in SIBYLLRun?

afedynitch replied:
Ok

HDembinski commented:
`SIBYLL*` is an unfortunate naming choice for us. I agree with @jncots that CAP style for the new classes is the way to be PEP8 compliant, see https://peps.python.org/pep-0008/#class-names

afedynitch commented:
@HDembinski I fixed the issues that you pointed out.

Further, I updated Sibyll* to patch release 2 that came in today with a bugfix. After your comments are resolved, I would like to merge this with to_release branch instead of main to bundle the numerous open PRs.

jncots commented on cmake_ext.py:
```diff
@@ -53,6 +53,7 @@ def get_models():
     # sib23c02
     # sib23c03
     # dev_dpmjetIII193=/full/path/to/dir/dpmjetIII-19.3
+    # dev_sib23d=/full/path/to/dir/sibyll
```
This should be removed because sib23d_start is added.

jncots commented on setup.py:
```diff
@@ -17,9 +17,9 @@
-# for convenience, support building extra models via extra.cfg
-# extra.cfg is not tracked by git, so can be freely modified
-# extra.cfg example:
+# for convenience, support building extra models via model.cfg
```
It is `models.cfg`. More complete comment is in `get_models` function. Here it is better to write something like "Refer to the comments for 'get_models' on instructions for adding extra models."

afedynitch replied:
OK, changed

jncots commented on tests/test_cli.py:
```diff
@@ -195,7 +196,7 @@ def test_model_3():
         stderr=(
             "Error: model=sib is ambiguous, matches (SIBYLL-2.1, SIBYLL-2.3, "
-            "SIBYLL-2.3c, SIBYLL-2.3d)"
+            "SIBYLL-2.3Star-mix, SIBYLL-2.3c, SIBYLL-2.3d)"
```
Maybe "SIBYLL-2.3Star-Mix"?

afedynitch replied:
This is just the name how it will appear in labels. I prefer the dash here

jncots reviewed (commented):
It is great to have a new model. I have found only a couple of minor suggestions for improvement.

afedynitch commented:
I implemented the comments by @jncots.

Also, I fixed the cross section readout for nuclei. It was necessary to include more categories into `CrossSectionData` to accommodate the special categories for nuclei, such as "production" and "quasielastic". I updated the tests to accomodate these changes.

jncots commented on src/chromo/common.py:
```diff
@@ -36,55 +36,72 @@
+        Cross section for pure elastic scattering of incoming particle
         without any new particle generation.
+    quasielastic : float
+        Nuclear cross section defined as total minus quasielastic.
```
It is confusing to name `quasielastic` if it is total-quasielastic. Maybe sort of compl2quasiel - complimentary to quasielastic?

afedynitch replied:
I improved the docstring.

jncots commented on src/chromo/common.py:
```diff
@@ -36,55 +36,72 @@
     inelastic: float = np.nan
     elastic: float = np.nan
+    prod: float = np.nan
+    quasielastic: float = np.nan
```
Could you add the description for `prod`?

afedynitch replied:
Added.

jncots commented on src/chromo/common.py:
```diff
@@ -36,55 +36,72 @@
+        if float(self.diffractive_sum) > 0.0:
+            return self.diffractive_sum
+        return sum(
+            [
```
Better to use np.nansum

afedynitch replied:
Good idea!

jncots commented on examples/cross_section.ipynb:
It would be good to add plt.ylabel("$\sigma$, mbarn")

jncots commented on tests/test_dpmjetIII.py:
```diff
@@ -39,3 +50,35 @@ def run_cross_section_ntrials():
+    assert_allclose(c.total, 38.9, atol=0.1)
+    assert_allclose(c.inelastic, 32.1, atol=0.1)
+    assert_allclose(c.elastic, 6.9, atol=0.1)
+    assert_allclose(
```
If 38.9, 32.1 and 6.9 are some kind of constants, maybe it is better to give them names like `total_at_10GeV` and so on.

jncots replied:
Or leave a comment that those numbers are for specific kinematics

afedynitch replied:
I addeed a comment, but it's just trivial. Simply what DPMJET returns for this calls today and not related to any data reference.

jncots commented on tests/test_dpmjetIII.py:
```diff
@@ -39,3 +50,35 @@ def run_cross_section_ntrials():
+    assert_allclose(c.elastic, 118.1, atol=0.1)
+    assert_allclose(c.prod, 298.7, atol=0.1)
+    assert_allclose(c.quasielastic, 144.6, atol=0.1)
+    assert_allclose(
```
The same is here

jncots reviewed (commented):
Great job, a lot of changes. Just a few fixes are needed.

jncots reviewed (approved):
Ready for merge
